src/librustc/traits/error_reporting/suggestions.rs

```diff
@@ -1699,20 +1715,41 @@ impl<'v> Visitor<'v> for ReturnsVisitor<'v> {
     }
 
     fn visit_expr(&mut self, ex: &'v hir::Expr<'v>) {
-        if let hir::ExprKind::Ret(Some(ex)) = ex.kind {
-            self.0.push(ex);
+        match ex.kind {
```

Comment:
The changes here feel unclear to me re. what is being achieves. Could you leave more comments?

Reply:
All of this is to be able to point at the leafs of the return path graph, not just at the tail expr (which can be match/if-else).